### DISCUSSION TITLE
Custom functions in micropython

### DISCUSSION THREAD

**Discussion by Redstonetill** [pybricks/support#1791] [Q&A] [answered]
I have followed the ''CONTRIBUTING.md" from pybricks-micropython and was able to build the firmware with new vars include in the 'pb_type_primehub.c'. I added vars like 'test' with the value of "Hello World" and it could be access through "PrimeHub().test". But now I want to take it to a practical level (the whole reason I am trying to build "my own" firmware is, because I need a algorithem to run way faster than "python" can, so I wanted to code it in C). So right now I have a problem with adding a custom function to the Prime Hub File. If anyone could help me with it, I would be very grateful since I have been working on this for 1.5 days : )

**Comment by dlech** [comment]
Can you share the function you need to optimize? Maybe there is a way to still do it in Python and meet the timing requirements.

Also, there are some other options to use native code without compiling your own firmware. For example https://docs.micropython.org/en/latest/develop/natmod.html is supported on SPIKE Prime.

But if you still want to learn how to modify the firmware, I would suggest starting with adding a function to the tools module (e.g. see `pb_module_tools_wait_obj` as an example). And then consider creating your own module instead of modifying the existing ones.

https://github.com/pybricks/pybricks-micropython/blob/master/pybricks/tools/pb_module_tools.c

**Reply by Redstonetill** [reply, marked as answer]
Since it is a search algorithem (A*) on f...ing huge graph there isn't any other way but to code it in C, but the approach with the "pb_module_tools.c" really help, and I have a simple function that just adds to numbers. This is the code I have added to the "pb_module_tools.c" if anyone would need it in the future:

This code at line 347
```c
static mp_obj_t pb_module_tools_add_ints(mp_obj_t a_obj, mp_obj_t b_obj) {
   int a = mp_obj_get_int(a_obj);
   int b = mp_obj_get_int(b_obj);
   return mp_obj_new_int(a + b);
}
static MP_DEFINE_CONST_FUN_OBJ_2(pb_module_tools_add_ints_obj, pb_module_tools_add_ints);
```

And this code into the "tools_globals_table[]" Array (the array is now at 360):
```c
{ MP_ROM_QSTR(MP_QSTR_add),        MP_ROM_PTR(&pb_module_tools_add_ints_obj)         },
```

**Reply by Redstonetill** [reply]
![grafik](https://github.com/user-attachments/assets/88190d3e-71de-497b-a9f3-dbacabede7d0)
![grafik](https://github.com/user-attachments/assets/3d8609dc-b26e-48d6-acc9-52cd0621e493)
(Code was uploaded to the Spike Prime Hub trough pybricksdev)

**Reply by dlech** [reply]
Glad you got it figured out.

And just to be sure in case it wasn't clear, the native module option would let you code functions in C similar to this, but you don't have to build custom firmware or re-flash each time you change your code.

**Reply by Redstonetill** [reply]
Ok thanks for the heads up :) But I think that I am going to do it over the more complicated way, because I think that it will look a lot better (I am doing this for a competition) in the competition if we can say that we (my team and I) are using an customized firmware based on pybricks and by programming the algorithm on my desktop (where the code will be executed in no time) the porting onto the spike prime hopefully shouldnt take to long. But still thanks for information maybe I will need it going down the line, you never know :)

**Reply by dlech** [reply]
Good luck with the competition.

We just like to make sure everyone knows all of the options so they can make the best decision. :smile:
